lib/grpc_provider.cpp

```diff
@@ -0,0 +1,284 @@
+                const auto& registerRet = providerMsg.registerret();
+                if (registerRet.err() != Core::ErrorNumber::OK) {
+                    throw std::runtime_error("Register fail!");
+                }
+            }
+
+            case ProviderMsg::kStartGameArgs: {
+                const auto& startGameArgs = providerMsg.startgameargs();
+                const auto& grpcUids = startGameArgs.userid();
+                std::vector<int> uids{grpcUids.begin(), grpcUids.end()};
+                auto roomid = startGameArgs.roomid();
+                mGameMap.emplace(std::make_pair(
+                    roomid, GameAndCardPool(roomid, uids, this)));
+            }
+
+            case ProviderMsg::kCloseGameArgs: {
+                throw std::runtime_error("unimplemented CloseGameArgs!");
+            }
+
+            case ProviderMsg::kQueryStateArgs: {
+                throw std::runtime_error("unimplemented QueryStateArgs!");
+            }
+
+            case ProviderMsg::kUserOperationArgs: {
+                const auto& userOperationArgs = providerMsg.useroperationargs();
+                ExplodingKittensProto::UserOperation userOperation;
+                userOperationArgs.custom().UnpackTo(&userOperation);
+                HandleUserMsg(userOperationArgs.roomid(),
+                              userOperationArgs.userid(), userOperation);
+            }
+        }
+    }
+}
+
+void Provider::SendStartGame(int roomid, int uid,
+                             const std::vector<CardType>& cards,
+                             const std::vector<int>& uids) {
+    Core::ProviderMsg startGameMsg;
+    auto* startGameArgs = startGameMsg.mutable_notifymsgargs();
+    startGameArgs->set_roomid(roomid);
+    startGameArgs->set_userid(uid);
+    ExplodingKittensProto::NotifyMsg notifyStartGameMsg;
+
+    ExplodingKittensProto::GameStart* gameStart =
+        notifyStartGameMsg.mutable_gamestart();
+    gameStart->set_userid(uid);
+    *(gameStart->mutable_userids()) =
+        google::protobuf::RepeatedField<google::protobuf::int32>(uids.begin(),
+                                                                 uids.end());
+    *(gameStart->mutable_cards()) =
+        google::protobuf::RepeatedField<google::protobuf::int32>(cards.begin(),
+                                                                 cards.end());
+
+    startGameArgs->mutable_custom()->PackFrom(notifyStartGameMsg);
+
+    mStream->Write(startGameMsg);
+}
+
+void Provider::SendRoundStart(int roomid, int uid,
+                              const std::vector<int>& uids) {
+    Core::ProviderMsg roundStartMsg;
+    auto* roundStartArgs = roundStartMsg.mutable_notifymsgargs();
+    roundStartArgs->set_roomid(roomid);
+    ExplodingKittensProto::NotifyMsg notifyRoundStartMsg;
+
+    ExplodingKittensProto::RoundStart* roundStart =
+        notifyRoundStartMsg.mutable_roundstart();
+    roundStart->set_userid(uid);
+
+    for (auto id : uids) {
```

Comment:
I think int is better than auto here.
1. const auto& is senseless,  auto is more efficient.
2. const auto looks odd
3. auto is not clear, because perhaps it is an iterator or a built-in type.

Reply:
fixed